node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,379 @@
+    ctx: &mut Context,
+    peerid: PeerId,
+    rep: ReputationChange,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    ctx.send_message(AllMessages::NetworkBridge(
+        NetworkBridgeMessage::ReportPeer(peerid, rep),
+    ))
+    .await
+}
+
+/// Distribute a given valid bitfield message.
+///
+/// Can be originated by another subsystem or received via network from another peer.
+async fn distribute<Context>(
+    ctx: &mut Context,
+    tracker: &mut Tracker,
+    message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    let BitfieldGossipMessage {
+        relay_parent,
+        signed_availability,
+    } = message;
+    // concurrently pass on the bitfield distribution to all interested peers
+    let interested_peers = tracker
+        .peer_views
+        .iter()
+        .filter_map(|(peerid, view)| {
+            if view.contains(&relay_parent) {
+                Some(peerid.clone())
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<PeerId>>();
+
+    let message = BitfieldGossipMessage {
+        relay_parent,
+        signed_availability,
+    };
+    let bytes = message.encode();
+    ctx.send_message(AllMessages::NetworkBridge(
+        NetworkBridgeMessage::SendMessage(
+            interested_peers,
+            BitfieldDistribution::PROTOCOL_ID,
+            bytes,
+        ),
+    ))
+    .await?;
+    Ok(())
+}
+
+/// Handle an incoming message from a peer.
+async fn process_incoming_peer_message<Context>(
+    ctx: &mut Context,
+    tracker: &mut Tracker,
+    peerid: PeerId,
+    message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+    Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+{
+    // we don't care about this, not part of our view
+    if !tracker.view.contains(&message.relay_parent) {
+        return modify_reputiation(ctx, peerid, COST_NOT_INTERESTED).await;
+    }
+
+    // Ignore anything the overseer did not tell this subsystem to work on
+    let mut job_data = tracker.per_relay_parent.get_mut(&message.relay_parent);
+    let job_data: &mut _ = if let Some(ref mut job_data) = job_data {
+        job_data
+    } else {
+        return modify_reputiation(ctx, peerid, COST_NOT_INTERESTED).await;
+    };
+
+    let validator_set = &job_data.validator_set;
+    if validator_set.len() == 0 {
+        return modify_reputiation(ctx, peerid, COST_MISSING_PEER_SESSION_KEY).await;
+    }
+
+    // check all validators that could have signed this message
+    // @todo there must be a better way figuring this out cheaply
+    let signing_context = job_data.signing_context.clone();
+    if let Some(validator) = validator_set.iter().find(|validator| {
+        message
+            .signed_availability
+            .check_signature(&signing_context, validator)
```

Comment:
@coriolinus Introduced a `Signed` type that we could use for this: so this subsystem should distribute `Signed<Bitfield>` or something like that where we send that over the wire. The `Signed` type carries around a `ValidatorIndex` which is the index into the validator set. Of course, this is untrusted data coming from the network so we need to do checked indexing into the validator set, but that should do the trick.